ltq-dsl-base: fix line uptime

alanswanson opened the pull request:
The Lantiq firmware under reports the uptime due to incorrect ticks per second bug - 1000 used in firmware calculation versus 1024 in modem hardware timer. Apply workaround. Both modem clock and system clock will still drift independently but now at the usual few seconds per day.

Original discussion at https://forum.openwrt.org/t/lantiq-dsl-modem-line-uptime-drift-lagging/ and seems unlikely firmware bug will ever be fixed.

Since there is another ltq-dsl-base pull request https://github.com/openwrt/openwrt/pull/3660 and other [patches](https://lists.infradead.org/pipermail/openwrt-devel/2020-December/032523.html) on the openwrt-devel mailing list, would be nice to get this added at the same time.

adschm commented:
Please bump PKG_RELEASE.

alanswanson commented:
The other ltq-dsl-base PR already bumps the version as does the email patches (in a separate patch/commit) so will conflict. If they're all hopefully merged around same should only need one (version bump)?

adschm commented:
> The other ltq-dsl-base PR already bumps the version as does the email patches (in a separate patch/commit) so will conflict. If they're all hopefully merged around same time should only need one (version bump)?

Each change should have it's own bump.

alanswanson committed:
ltq-dsl-base: fix line uptime

The Lantiq firmware under reports the uptime due to incorrect
ticks per second bug - 1000 used in firmware calculation versus
1024 in modem hardware timer. Apply workaround.

Both modem clock and system clock will still drift
independently but now at the usual few seconds per day.

https://forum.openwrt.org/t/lantiq-dsl-modem-line-uptime-drift-lagging/

Signed-off-by: Alan Swanson <reiver@improbabiity.net>

dhewg commented:
This fixes up the returned data from the kernel driver. Can't we instead fix the driver, so all users of this data are fixed?
And there're more time based values returned, aren't those affected too?

alanswanson commented:
That would be more sensible. Took a bit of grepping but looks to not be the firmware but **drv_dsl_cpe_api-4.17.18.6** at _src/include/drv_dsl_cpe_pm_core.h_ has two relevant defines that are used in timing and calculation of nElapsedTime and other time reporting (not affecting general DSL operation).

`#define DSL_PM_COUNTER_POLLING_CYCLE (1000)`
`#define DSL_PM_MSEC (1000)`

Changing DSL_PM_MSEC to 1024 would make nElapsedTime along with nCurr[Show|1Day|15Min|5min]Time worse, since it's used in division so would be doubly under reporting.
Changing DSL_PM_COUNTER_POLLING_CYCLE to 1024 would set msecTimeFrame as used in DSL_DRV_PM_SyncTimeUpdate for all time updates and fix the issue.

Not tested but a second opinion would be good.

dhewg commented:
From a quick glance, I don't see how changing DSL_PM_COUNTER_POLLING_CYCLE would make a difference, since, apart from the first run, DSL_DRV_PM_SyncTimeUpdate() calculates msecTimeFrame as the time difference since the last run.

Are you really sure the bug is due to a 1024/1000 mismatch? The involved time values all seem to be based on jiffies, see the whole DSL_DRV_PM_TIME_GET/DSL_DRV_TimeMSecGet/IFXOS_ElapsedTimeMSecGet mess. And 1000 sounds correct there.

But to me it looks like how elapsed time is measured here is just made of fail. Instead of saving a time stamp in DSL_DRV_PM_ElapsedTimeReset() and calculating the time difference when required, the time difference between consecutive timer runs is added... Maybe the overhead of that mess just *looks* like ~2.4%? :rofl:

alanswanson commented:
No, you're right on the DSL_PM_COUNTER_POLLING_CYCLE. The code is horrible! But the 1024/1000 difference just seems too much coincidence. 

Anyway, the latest emailed patches kill the _lantiq_dsl.sh_ script so I'll close this task and see what I can do with the timers.

dhewg commented:
So far I didn't yet see a bug that explains exactly that factor. And it doesn't look like the firmware is involved here. But it's easy to miss something or get lost in all the lantiq abstraction layers.

Anyway, a quick & dirty script:
```
#!/bin/sh

while true; do
	sys=`cat /proc/uptime | cut -d'.' -f1`
	dsl=`ubus call dsl metrics | jsonfilter -e \@.uptime`
	diff=$(($sys-$dsl))
	xdsl=$(($dsl*1024/1000))
	xdiff=$(($sys-$xdsl))
	echo "$sys;$dsl;$diff;$xdsl;$xdiff"
	sleep 60
done
```

So it outputs
`system uptime;dsl uptime;difference;dsl uptime*1024/1000;difference`

Which should mean the 3rd field keeps growing and the last should be constant if you're right, right?

```
364;154;210;157;207
424;213;211;218;206
484;271;213;277;207
544;330;214;337;207
604;389;215;398;206
664;447;217;457;207
724;506;218;518;206
784;565;219;578;206
845;624;221;638;207
905;687;218;703;202
973;746;227;763;210
1033;804;229;823;210
1093;863;230;883;210
1153;922;231;944;209
1213;980;233;1003;210
```

We can ignore +/-1, but guess where I produced some load on the box ;) And this is only for the first 15 minutes...

dhewg commented:
My set doesn't invalidate your PR though. You could reopen it so it's clear this isn't solved yet. There definitely is a bug wrt uptime reporting, we're just not sure yet what the proper fix is. When we do know you could push it to this PR.

Did you try my suggestion above on how to measure time differences? Do you want me to give it a try? Due to the metrics stuff I'm already deeper in this mess than I ever planned to /o\

alanswanson commented:
I'd closed as this would change package from ltq-dsl-base to ltq-ltq-vdsl-vr9 and I don't have any WIP commit yet and as it's a PR rather than a bug/issue. Can reopen and title when we have a patch.

Not done a recent OpenWRT build so don't have your most excellent ubus dslstat speed patches but backporting your test script (if anyone else wants to check);
_dsl=\`/etc/init.d/dsl_control lucistat | grep "dsl.line_uptime=" | awk -F "=" '{print $2}'\`_

Line uptime is currently 23days 12hours.
```
7773735;1983126;5790609;2030721;5743014 -< CONSISTENT
7773797;1983187;5790610;2030783;5743014
7773859;1983247;5790612;2030844;5743015
7773921;1983311;5790610;2030910;5743011 <- LOAD START
7773986;1983375;5790611;2030976;5743010
7774051;1983439;5790612;2031041;5743010 -> LOAD END
7774117;1983500;5790617;2031104;5743013
7774179;1983560;5790619;2031165;5743014
```

So, yes, does looks to be duff timer code!

dhewg commented:
FYI my set finally landed, in case you waited for that to continue to work on this uptime issue
